onnxruntime/core/session/device_allocator.cc

```diff
@@ -57,6 +57,70 @@ ORT_API_STATUS_IMPL(OrtApis::CreateAllocator, const OrtSession* sess, const OrtM
   API_IMPL_END
 }
 
+ORT_API_STATUS_IMPL(OrtApis::CreateAllocatorForSharing, _In_ const OrtMemoryInfo* mem_info,
+                    _In_ const OrtArenaCfg* arena_cfg, _Outptr_ OrtAllocator** out) {
+  using namespace onnxruntime;
+  if (!mem_info) {
+    return OrtApis::CreateStatus(ORT_INVALID_ARGUMENT, "OrtMemoryInfo is null");
+  }
+
+  // TODO should we allow sharing of non-CPU allocators?
+  if (mem_info->device.Type() != OrtDevice::CPU) {
+    return OrtApis::CreateStatus(ORT_INVALID_ARGUMENT, "Only CPU devices are supported for now.");
+  }
+
+  // determine if arena should be used
+  bool create_arena = mem_info->alloc_type == OrtArenaAllocator;
+
```

Comment:
Is it too big a change (or should be a separate PR) to move the info on whether an arena is involved or not out of OrtMemoryInfo? Whether the memory is coming from an arena or not seems orthogonal to the type of memory it is, and I would have expected OrtMemoryInfo to just be about the type of memory.

Reply:
Yeah, this will involve API changes as well. Should be tackled when we tackle allocator refactoring as a dedicated work item.